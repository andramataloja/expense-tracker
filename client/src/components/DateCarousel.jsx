import React, { useEffect, useState } from "react";
import Carousel from "@brainhubeu/react-carousel";
import axios from "axios";
import "@brainhubeu/react-carousel/lib/style.css";
import ArrowForwardIosRoundedIcon from "@material-ui/icons/ArrowForwardIosRounded";
import ArrowBackIosRoundedIcon from "@material-ui/icons/ArrowBackIosRounded";
import { useDispatch, useSelector } from "react-redux";
import { fetchMonth, fetchYear } from "../actions/actions";
import { makeStyles, Box } from "@material-ui/core";

const useStyles = makeStyles(theme => ({
  carousel: {
    transform: "none",
    width: "330px",
    marginTop: "10px"
  },
  customArrow: {
    color: theme.palette.primary.main,
    marginLeft: "10px",
    marginRight: "10px"
  },
  carouselSlides: {
    fontFamily: theme.typography.fontFamily
  }
}));

const DateCarousel = () => {
  const classes = useStyles();
  const dispatch = useDispatch();
  const monthState = useSelector(state => state.month);
  const yearState = useSelector(state => state.year);

  const [months, setMonths] = useState([]);
  const [years, setYears] = useState([]);

  useEffect(() => {
    axios
      .get("/get/months")
      .then(res => (res.data.length !== 0 ? setMonths(res.data) : null))
      .catch(err => console.log(err));
  }, []);

  useEffect(() => {
    let lastYear = new Date().getFullYear() - 1;
    let maxOffset = 10;

    const years = [];
    for (let x = 0; x <= maxOffset; x++) {
      years.push(lastYear + x);
    }
    setYears(years);
  }, [yearState]);

  return (
    <Box className={classes.carouselSlides} data-testid="box">
      <Carousel
        data-testid="yearState"
        className={classes.carousel}
        centered
        itemWidth={120}
        arrowRight={
          <ArrowForwardIosRoundedIcon className={classes.customArrow} />
        }
        arrowLeft={<ArrowBackIosRoundedIcon className={classes.customArrow} />}
        arrowLeftDisabled={<Box />}
        arrowRightDisabled={<Box />}
        addArrowClickHandler
        slidesPerPage={3}
        clickToChange
        value={years.indexOf(yearState)}
        onChange={value => dispatch(fetchYear(years[value]))}
        slides={years.map((year, index) => (
          <Box key={index}>{year}</Box>
        ))}
      ></Carousel>
      <Carousel
        data-testid="monthState"
        className={classes.carousel}
        centered
        itemWidth={120}
        arrowRight={
          <ArrowForwardIosRoundedIcon className={classes.customArrow} />
        }
        arrowLeft={<ArrowBackIosRoundedIcon className={classes.customArrow} />}
        arrowLeftDisabled={<Box />}
        arrowRightDisabled={<Box />}
        addArrowClickHandler
        slidesPerPage={3}
        clickToChange
        value={monthState}
        onChange={value => dispatch(fetchMonth(value))}
        slides={months.map(month => (
          <Box key={month.month_id}>{month.month}</Box>
        ))}
      ></Carousel>
    </Box>
  );
};
export default DateCarousel;
