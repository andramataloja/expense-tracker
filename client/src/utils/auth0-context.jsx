import React, { useState, useEffect, useContext } from "react";
import createAuth0Client from "@auth0/auth0-spa-js";

const DEFAULT_REDIRECT_CALLBACK = () =>
  window.history.replaceState({}, document.title, window.location.pathname);

export const Auth0Context = React.createContext();
export const useAuth0 = () => useContext(Auth0Context);
export const Auth0Provider = ({
  children,
  onRedirectCallback = DEFAULT_REDIRECT_CALLBACK
}) => {
  const [isAuthenticated, setIsAuthenticated] = useState();
  const [user, setUser] = useState();
  const [auth0Client, setAuth0Client] = useState();
  const [isLoading, setIsLoading] = useState(true);

  const config = {
    domain: process.env.REACT_APP_AUTH0_DOMAIN,
    client_id: process.env.REACT_APP_AUTH0_CLIENT_ID,
    redirect_uri: "http://localhost:3000/Dashboard"
  };

  useEffect(() => {
    const initializeAuth0 = async () => {
      const auth0FromHook = await createAuth0Client(config);
      setAuth0Client(auth0FromHook);

      if (
        window.location.search.includes("code=") &&
        window.location.search.includes("state=")
      ) {
        const { appState } = await auth0FromHook.handleRedirectCallback();
        onRedirectCallback(appState);
      }

      const isAuthenticated = await auth0FromHook.isAuthenticated();

      setIsAuthenticated(isAuthenticated);

      if (isAuthenticated) {
        const user = await auth0FromHook.getUser();
        setUser(user);
      }

      setIsLoading(false);
    };
    initializeAuth0();
  }, []);

  const handleRedirectCallback = async () => {
    setIsLoading(true);
    await auth0Client.handleRedirectCallback();
    const user = await auth0Client.getUser();
    setIsLoading(false);
    setIsAuthenticated(true);
    setUser(user);
  };
  return (
    <Auth0Context.Provider
      value={{
        isAuthenticated,
        user,
        isLoading,
        handleRedirectCallback,
        getIdTokenClaims: (...p) => auth0Client.getIdTokenClaims(...p),
        loginWithRedirect: (...p) => auth0Client.loginWithRedirect(...p),
        getTokenSilently: (...p) => auth0Client.getTokenSilently(...p),
        logout: (...p) => auth0Client.logout(...p)
      }}
    >
      {children}
    </Auth0Context.Provider>
  );
};
