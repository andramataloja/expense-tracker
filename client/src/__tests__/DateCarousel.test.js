import React from "react";
import {
  cleanup,
  render,
  fireEvent,
  screen,
  wait,
  waitForElement
} from "@testing-library/react";
import DateCarousel from "../components/DateCarousel";
import { useAuth0 } from "../utils/auth0-context";
import { Provider } from "react-redux";
import { act } from "react-dom/test-utils";

import thunk from "redux-thunk";
import configureMockStore from "redux-mock-store";
import mockAxios from "axios";

const API = "http://localhost:3000";

const user = {
  email: "johndoe@me.com",
  email_verified: true,
  sub: "google-oauth2|12345678901234"
};

jest.mock("../utils/auth0-context");
jest.mock("../store.js");

describe("DateCarousel", () => {
  const middlewares = [thunk];
  const mockStore = configureMockStore(middlewares);

  afterEach(cleanup);
  beforeEach(() => {
    useAuth0.mockReturnValue({
      isAuthenticated: true,
      user,
      logout: jest.fn(),
      loginWithRedirect: jest.fn()
    });
    require("mutationobserver-shim");
    mockAxios.get.mockImplementationOnce(() => Promise.resolve(mockData));
  });

  const mockData = {
    data: [
      {
        month_id: 1,
        month: "January",
        month_nr: 0
      },
      {
        month_id: 2,
        month: "February",
        month_nr: 1
      }
    ]
  };

  it("fetches successfully data from an API", async () => {
    expect(mockAxios.get(`${API}/get/months`)).resolves.toEqual(mockData);
    expect(mockAxios.get).toHaveBeenCalledWith(`${API}/get/months`);
  });

  it("should take a snapshot of datecarousel", async () => {
    const store = mockStore();
    const { asFragment } = render(
      <Provider store={store}>
        <DateCarousel />
      </Provider>
    );
    await act(wait);

    expect(asFragment(<DateCarousel />)).toMatchSnapshot();
  });

  /* it("renders", async () => {
    const store = mockStore();
    const { container } = render(
      <Provider store={store}>
        <DateCarousel />
      </Provider>
    );
    await act(wait);

    expect(container.firstChild).toHaveClass(
      "BrainhubCarousel__customArrows BrainhubCarousel__arrow--disable BrainhubCarousel__custom-arrowLeft"
    );
  }); */
});
