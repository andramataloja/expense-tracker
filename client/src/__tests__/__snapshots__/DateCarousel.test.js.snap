// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateCarousel should take a snapshot of datecarousel 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-5 makeStyles-carouselSlides-3"
    data-testid="box"
  >
    <div>
      <div
        class="BrainhubCarousel makeStyles-carousel-1"
      >
        <div
          class="BrainhubCarousel__customArrows BrainhubCarousel__arrow--disable BrainhubCarousel__custom-arrowLeft"
        >
          <div
            class="MuiBox-root MuiBox-root-6"
          />
        </div>
        <div
          class="BrainhubCarousel__trackContainer"
        >
          <ul
            class="BrainhubCarousel__track BrainhubCarousel__track--draggable"
            style="width: 0px; transform: translateX(-60px); margin-left: 0px;"
          >
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable BrainhubCarouselItem--active"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-18"
              >
                2019
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-19"
              >
                2020
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-20"
              >
                2021
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-21"
              >
                2022
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-22"
              >
                2023
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-23"
              >
                2024
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-24"
              >
                2025
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-25"
              >
                2026
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-26"
              >
                2027
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-27"
              >
                2028
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-28"
              >
                2029
              </div>
            </li>
          </ul>
        </div>
        <div
          class="BrainhubCarousel__customArrows BrainhubCarousel__custom-arrowRight"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-customArrow-2"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.38 21.01c.49.49 1.28.49 1.77 0l8.31-8.31c.39-.39.39-1.02 0-1.41L9.15 2.98c-.49-.49-1.28-.49-1.77 0s-.49 1.28 0 1.77L14.62 12l-7.25 7.25c-.48.48-.48 1.28.01 1.76z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div>
      <div
        class="BrainhubCarousel makeStyles-carousel-1"
      >
        <div
          class="BrainhubCarousel__customArrows BrainhubCarousel__arrow--disable BrainhubCarousel__custom-arrowLeft"
        >
          <div
            class="MuiBox-root MuiBox-root-8"
          />
        </div>
        <div
          class="BrainhubCarousel__trackContainer"
        >
          <ul
            class="BrainhubCarousel__track BrainhubCarousel__track--draggable"
            style="width: 0px; margin-left: 0px; transform: translateX(-60px);"
          >
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable BrainhubCarouselItem--active"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-29"
              >
                January
              </div>
            </li>
            <li
              class="BrainhubCarouselItem BrainhubCarouselItem--clickable"
              style="padding-right: 0px; padding-left: 0px; width: 120px; max-width: 120px; min-width: 120px;"
            >
              <div
                class="MuiBox-root MuiBox-root-30"
              >
                February
              </div>
            </li>
          </ul>
        </div>
        <div
          class="BrainhubCarousel__customArrows BrainhubCarousel__custom-arrowRight"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root makeStyles-customArrow-2"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M7.38 21.01c.49.49 1.28.49 1.77 0l8.31-8.31c.39-.39.39-1.02 0-1.41L9.15 2.98c-.49-.49-1.28-.49-1.77 0s-.49 1.28 0 1.77L14.62 12l-7.25 7.25c-.48.48-.48 1.28.01 1.76z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
